Fix eslint

adamwoodnz opened the pull request:
Fixes #137 

Adds the `@wordpress/eslint-plugin/recommended` settings which includes Prettier formatting. See the commits for the steps I've taken to fix the 396 problems originally identified.

After auto fixing eslint/prettier problems, and the trivial fixes, there are 8 errors remaining which are likely to cause bugs:

```
$ wp-scripts lint-js

/home/runner/work/wporg-two-factor/wporg-two-factor/settings/src/components/email-address.js
Error:   48:4  error  Unexpected alert        no-alert
Error:   48:4  error  'alert' is not defined  no-undef

/home/runner/work/wporg-two-factor/wporg-two-factor/settings/src/components/totp.js
Error:   144:23  error  React Hook "useCallback" is called conditionally. React Hooks must be called in the exact same order in every component render                                                                 react-hooks/rules-of-hooks
Error:   182:23  error  React Hook "useCallback" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return?  react-hooks/rules-of-hooks

/home/runner/work/wporg-two-factor/wporg-two-factor/settings/src/script.js
Error:    77:3  error  'history' is not defined  no-undef
Error:   11[9](https://github.com/WordPress/wporg-two-factor/actions/runs/4825873555/jobs/8597161531?pr=138#step:5:10):4  error  'history' is not defined  no-undef

/home/runner/work/wporg-two-factor/wporg-two-factor/settings/src/utilities.js
Error:   [10](https://github.com/WordPress/wporg-two-factor/actions/runs/4825873555/jobs/8597161531?pr=138#step:5:11):21  error  React Hook "useEntityRecord" is called in function "getUserRecord" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter. React Hook names must start with the word "use"  react-hooks/rules-of-hooks
Error:   [14](https://github.com/WordPress/wporg-two-factor/actions/runs/4825873555/jobs/8597161531?pr=138#step:5:15):[25](https://github.com/WordPress/wporg-two-factor/actions/runs/4825873555/jobs/8597161531?pr=138#step:5:26)  error  React Hook "useSelect" is called in function "getUserRecord" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter. React Hook names must start with the word "use"        react-hooks/rules-of-hooks

✖ 8 problems (8 errors, 0 warnings)
```

## How to test

Regression test the app. The 8 non-trivial errors were related to saving email address and switching between QR code and manual totp setup methods.

adamwoodnz committed:
Add eslint config

adamwoodnz committed:
Add top level npm script for linting settings module

adamwoodnz committed:
Fix all auto-fixable eslint errors

Fixed with `yarn lint:js --fix`

adamwoodnz committed:
Add github lint workflow

adamwoodnz committed:
Disable JSDoc param type eslint errors

adamwoodnz committed:
Fix trivial errors

adamwoodnz committed:
Update generated params to use `props` rather than `root0`

adamwoodnz commented on .github/workflows/lint.yml:
```diff
@@ -0,0 +1,31 @@
+# This workflow is triggered on pushes to trunk, and any PRs.
+on:
+  push:
+    branches: [fix/137-eslint]
```
Update this once we're ready to run on all branches

adamwoodnz replied:
Done

dd32 commented on settings/src/components/email-address.js:
```diff
@@ -18,23 +14,24 @@ import { GlobalContext } from '../script';
 export default function EmailAddress() {
-	const { userRecord }                                           = useContext( GlobalContext );
+	const { userRecord } = useContext( GlobalContext );
 	const { record, edit, save, editedRecord, hasEdits, isSaving } = userRecord;
-	const [ emailError, setEmailError ]                            = useState( '' );
-	const [ justChangedEmail, setJustChangedEmail ]                = useState( false );
+	const [ emailError, setEmailError ] = useState( '' );
+	const [ justChangedEmail, setJustChangedEmail ] = useState( false );
```
Is it possible for us to avoid these whitespace changes? 

I get that JS and PHP coding standards don't match, but aligning variable assignments makes for so much more readable code..

adamwoodnz replied:
Sure I'll see if we can enforce that

adamwoodnz replied:
It seems that aligning variable assignments isn't possible with Prettier, which is included by default in `@wordpress/eslint-plugin/recommended`. Closest issue I've found is https://github.com/prettier/prettier/issues/1622

Personally I'm used to this formatting, but I know it irks others. If we feel really strongly we could use Prettier to get things in the shape we want in this PR, then disable it in our eslint configuration.

```javascript
module.exports = {
	extends: 'plugin:@wordpress/eslint-plugin/recommended',

	rules: {
		...,
		'prettier/prettier': 0,
	},
};
```

Essentially the current state, but I'll revert these particular whitespace changes.

adamwoodnz replied:
Discussed elsewhere: the team favours following Gutenberg standards and keeping Prettier enabled.

adamwoodnz committed:
Increase formatting line length to 100 chars via prettier

StevenDufresne commented on settings/src/edit.js:
```diff
@@ -10,11 +10,13 @@ import { Placeholder } from '@wordpress/components';
 				label={ __( 'Two-Factor Authentication', 'wporg' ) }
-			> </Placeholder>
+			>
+				{ ' ' }
```
This should be removed.

adamwoodnz replied:
Done

adamwoodnz committed:
Fix rules of hooks violations for getUserRecord

Stop calling useSelect conditionally, and use naming convention for custom hook.

adamwoodnz commented on settings/src/utilities.js:
```diff
@@ -3,14 +3,15 @@ import { store as coreDataStore, useEntityRecord } from '@wordpress/core-data';
 
-	// Polyfill in isSaving.
-	if ( undefined === userRecord.isSaving ) {
-		userRecord.isSaving = useSelect( ( select ) => select( coreDataStore ).isSavingEntityRecord( 'root', 'user', userId ) );
-	}
+	userRecord.isSaving = useSelect( ( select ) =>
+		select( coreDataStore ).isSavingEntityRecord( 'root', 'user', userId )
+	);
```
Calling hooks conditionally can have unpredictable results, and I'm not seeing any changes in behaviour on the frontend when calling this unconditionally. I believe keeping this value in sync with the store at all times is desirable. @dd32 @iandunn is there a reason this has been written this way? Is it intended to be a perf optimization?

iandunn replied:
It looks like it was added in 2925a2f66597b3bb7f07e4a54c1c7d85e9032285. I'm not sure why, but if you're not seeing any problems then it seems safe to go ahead and change it. We can always dig deeper if we notice a problem in the future 👍🏻

adamwoodnz committed:
Fix rules of hooks in SetupMethod

Stop conditional calling of useCallback hooks by creating 2 smaller simpler components

adamwoodnz committed:
Remove unnecessary placeholder string

adamwoodnz committed:
Fix eslint undefined `history` errors

adamwoodnz committed:
Replace `alert` usage with email error notice

adamwoodnz commented on settings/src/components/email-address.js:
```diff
@@ -47,44 +43,43 @@ export default function EmailAddress() {
-		} catch( error ) {
-			alert( error.message );
+		} catch ( error ) {
+			setEmailError( error.message );
```
I'm unsure whether this is an appropriate message to be displayed to a user or what it'll look like. Needs testing.

adamwoodnz replied:
Tested this by returning a 404 and it appears it'll handle it appropriately

![Screen Shot 2023-05-01 at 4 19 18 PM](https://user-images.githubusercontent.com/1017872/235406067-5725447b-06f9-4cbe-b752-feab1f3574d3.jpg)

adamwoodnz commented on settings/.eslintrc.js:
```diff
@@ -0,0 +1,14 @@
+	extends: 'plugin:@wordpress/eslint-plugin/recommended',
+
+	rules: {
+		'jsdoc/require-param-type': 0,
```
This disables the requirement to have types declared on all function parameters. I've decided to disable for now as it would be a significant effort. We could add these at a later date to better document what's expected.

adamwoodnz commented on settings/.eslintrc.js:
```diff
@@ -0,0 +1,14 @@
+			'error',
+			{
+				...require( '@wordpress/prettier-config' ),
+				printWidth: 100,
```
Default is 80, which resulted in some of the longer strings having `{' '}` inserted when React components are inline within the text. 100 seems like a reasonable middle ground where this is rare.

adamwoodnz committed:
Update formatting after rebase

adamwoodnz committed:
Fix regression in setting url parameters

dd32 reviewed (approved):
While I might not agree 100% with some of the formats chosen by the JS formatter, I agree that maintaining style with Gutenberg and other WP JS projects is worth the tradeoffs.

Having some kind of standard is better than none.

adamwoodnz committed:
Update lint workflow to run on `push` to `trunk`, and remove JS build
